File: Algorithms/0633_Sum_of_Square_Numbers/Java/SumofSquareNumbersSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public boolean judgeSquareSum(int c) {
        if (c <= 2) return true;
        int temp = (int) Math.pow(c, 0.5) + 1;
        int[] cache = new int[temp];
        for (int i = 0; i < cache.length; i++) {
            cache[i] = i;
        }
        int left = 0, right = cache.length - 1;
        while (left <= right) {
            if (left * left + right * right == c) return true;
            if (left * left + right * right > c) {
                right--;
            } else if (left * left + right * right < c) {
                left++;
            }
        }
        return false;
    }
}

File: Algorithms/0633_Sum_of_Square_Numbers/Python/Sum_of_Square_Numbers_Solution_1.py

# Space: O(n)
# Time: O(n)

import math


class Solution:
    def judgeSquareSum(self, c):
        if c <= 2: return True
        cache = [i for i in range(math.ceil(pow(c, 0.5)) + 1)]
        left, right = 0, len(cache)-1
        while left <= right:
            if pow(cache[left], 2) + pow(cache[right], 2) == c: return True
            if pow(cache[left], 2) + pow(cache[right], 2) < c:
                left += 1
            elif pow(cache[left], 2) + pow(cache[right], 2) > c:
                right -= 1
        return False




